pallets/nft-marketplace/src/lib.rs

```diff
@@ -0,0 +1,330 @@
+			T::NonFungibles::transfer(&class_id.into(), &instance_id.into(), &Self::account())
+				.map_err(|_| Error::<T>::FailedNftTransfer)?;
+
+			// Put the nft for sale
+			let sale = Sale {
+				seller,
+				price: AskingPrice { currency, amount },
+			};
+			<Gallery<T>>::insert(class_id, instance_id, sale.clone());
+			Self::deposit_event(Event::ForSale(sale));
+
+			Ok(())
+		}
+
+		/// Remove an NFT from sale
+		///
+		/// The sellers of an NFT that is for sale can call this extrinsic to reclaim
+		/// ownership over their NFT and remove it from sale.
+		///
+		/// Fails if
+		///   - the nft is not for sale
+		///   - `origin` is not the seller of the NFT
+		///   - transferring the ownership of the NFT back to the seller fails
+		#[pallet::weight(10_000_000)]
+		pub fn remove(
+			origin: OriginFor<T>,
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+			let sale = <Gallery<T>>::get(class_id, instance_id).ok_or(Error::<T>::NotForSale)?;
+
+			// Ensure that the buyer is not the seller of the NFT
+			ensure!(who == sale.seller, Error::<T>::NotOwner);
+
+			// Transfer the NFT back to the seller, i.e., the original owner of this NFT
+			T::NonFungibles::transfer(&class_id.into(), &instance_id.into(), &sale.seller)
+				.map_err(|_| Error::<T>::FailedNftTransfer)?;
+
+			// Remove the NFT from the gallery
+			<Gallery<T>>::remove(class_id, instance_id);
+			Self::deposit_event(Event::Removed {
+				class_id,
+				instance_id,
+			});
+
+			Ok(())
+		}
+
+		/// Buy the given nft
+		///
+		/// Fails if
+		///   - the NFT is not for sale
+		///   - `origin` is the seller of the NFT
+		///   - `origin` does not have enough balance of the currency the nft is being sold in
+		///   - transferring the asking price from the buyer to the seller fails
+		///   - transferring the nft to the buyer fails
+		#[pallet::weight(10_000_000)]
+		#[transactional]
+		pub fn buy(
+			origin: OriginFor<T>,
+			class_id: T::ClassId,
+			instance_id: T::InstanceId,
+		) -> DispatchResult {
+			let buyer = ensure_signed(origin.clone())?;
+			let sale = <Gallery<T>>::get(class_id, instance_id).ok_or(Error::<T>::NotForSale)?;
+
+			// Ensure that the buyer is not the seller of the NFT
+			ensure!(buyer != sale.seller, Error::<T>::IsSeller);
+
+			// Make sure the buyer can pay for the NFT
+			T::Fungibles::can_withdraw(sale.price.currency, &buyer, sale.price.amount)
+				.into_result()
+				.map_err(|_| Error::<T>::InsufficientBalance)?;
+
+			// Have the buyer pay the seller for the NFT
+			T::Fungibles::transfer(
+				sale.price.currency,
+				&buyer,
+				&sale.seller,
+				sale.price.amount,
+				true,
+			)
+			.map_err(|_| Error::<T>::PaymentFailed)?;
+
+			// Transfer the NFT to the buyer
+			T::NonFungibles::transfer(
+				// Self::origin(),
+				&class_id.into(),
+				&instance_id.into(),
+				&buyer.clone(),
+			)
+			.map_err(|_| Error::<T>::FailedNftTransfer)?;
```

Comment:
As with the currency transfer, IMO it's better to let the underlying error be what we return so there's as much debugging information as possible in the block